File: dev/oracle/engine.py
"""Reference port of Sources/BackgammonEngine, used as a dev-only oracle.

This is a line-for-line translation of the Swift rules engine into Python so the
correctness-critical move-generation logic can be exercised with property tests
and fixtures in an environment without a Swift toolchain. It is NOT shipped in
the app. If you change the Swift engine, mirror the change here and re-run
`python3 dev/oracle/test_engine.py`.

Conventions match the Swift code exactly:
- points: list of length 25, index 0 unused; +n = n White, -n = n Black.
- bar/off: dict {1: white_count, -1: black_count} keyed by player sign.
- Internal generation runs on a normalized board (mover positive, high -> low).
"""
from __future__ import annotations
from dataclasses import dataclass
from typing import List, Optional

WHITE = 1
BLACK = -1


def opponent(player: int) -> int:
    return -player


@dataclass(frozen=True)
class Board:
    points: tuple            # length 25
    bar_white: int = 0
    bar_black: int = 0
    off_white: int = 0
    off_black: int = 0

    @staticmethod
    def starting() -> "Board":
        pts = [0,
               -2, 0, 0, 0, 0, 5,
               0, 3, 0, 0, 0, -5,
               5, 0, 0, 0, -3, 0,
               -5, 0, 0, 0, 0, 2]
        return Board(tuple(pts))

    def bar(self, player: int) -> int:
        return self.bar_white if player == WHITE else self.bar_black

    def off(self, player: int) -> int:
        return self.off_white if player == WHITE else self.off_black

    def owns(self, p: int, player: int) -> bool:
        return self.points[p] > 0 if player == WHITE else self.points[p] < 0

    def count(self, p: int, player: int) -> int:
        v = self.points[p]
        if player == WHITE:
            return v if v > 0 else 0
        return -v if v < 0 else 0

    def checker_count(self, player: int) -> int:
        c = self.bar(player) + self.off(player)
        for p in range(1, 25):
            c += self.count(p, player)
        return c

    def mirrored(self) -> "Board":
        np = [0] * 25
        for p in range(1, 25):
            np[p] = -self.points[25 - p]
        return Board(tuple(np), self.bar_black, self.bar_white,
                     self.off_black, self.off_white)

    def normalized(self, player: int) -> "Board":
        return self if player == WHITE else self.mirrored()

    @property
    def winner(self) -> Optional[int]:
        if self.off_white == 15:
            return WHITE
        if self.off_black == 15:
            return BLACK
        return None

    def win_type(self) -> Optional[str]:
        w = self.winner
        if w is None:
            return None
        loser = opponent(w)
        if self.off(loser) > 0:
            return "single"
        if self.bar(loser) > 0:
            return "backgammon"
        home = range(1, 7) if w == WHITE else range(19, 25)
        for p in home:
            if self.owns(p, loser):
                return "backgammon"
        return "gammon"

    def pip_count(self, player: int) -> int:
        pips = self.bar(player) * 25
        for p in range(1, 25):
            if self.owns(p, player):
                dist = p if player == WHITE else (25 - p)
                pips += self.count(p, player) * dist
        return pips


@dataclass(frozen=True)
class CheckerMove:
    frm: int
    to: int
    die: int


@dataclass(frozen=True)
class Play:
    moves: tuple            # tuple of CheckerMove
    result: Board


# --- normalized helpers (mover positive) ---

def can_land(b: Board, p: int) -> bool:
    return b.points[p] >= -1


def all_home(b: Board) -> bool:
    if b.bar_white > 0:
        return False
    for p in range(7, 25):
        if b.points[p] > 0:
            return False
    return True


def none_higher(b: Board, p: int) -> bool:
    if p >= 6:
        return True
    for q in range(p + 1, 7):
        if b.points[q] > 0:
            return False
    return True


def single_step_moves(b: Board, die: int) -> List[CheckerMove]:
    moves: List[CheckerMove] = []
    if b.bar_white > 0:
        entry = 25 - die
        if can_land(b, entry):
            moves.append(CheckerMove(25, entry, die))
        return moves
    for p in range(24, 0, -1):
        if b.points[p] > 0:
            dest = p - die
            if dest >= 1 and can_land(b, dest):
                moves.append(CheckerMove(p, dest, die))
    if all_home(b):
        for p in range(1, 7):
            if b.points[p] > 0:
                if p == die:
                    moves.append(CheckerMove(p, 0, die))
                elif die > p and none_higher(b, p):
                    moves.append(CheckerMove(p, 0, die))
    return moves


def apply_normalized(m: CheckerMove, board: Board) -> Board:
    pts = list(board.points)
    bw, bb = board.bar_white, board.bar_black
    ow = board.off_white
    if m.frm == 25:
        bw -= 1
    else:
        pts[m.frm] -= 1
    if m.to == 0:
        ow += 1
    else:
        if pts[m.to] < 0:
            pts[m.to] = 0
            bb += 1
        pts[m.to] += 1
    return Board(tuple(pts), bw, bb, ow, board.off_black)


def _explore(board: Board, remaining, acc, out: dict, max_used: list):
    if len(acc) > max_used[0]:
        max_used[0] = len(acc)
    if acc:
        out[(tuple(acc), board)] = Play(tuple(acc), board)
    if not remaining:
        return
    die = remaining[0]
    rest = remaining[1:]
    for mv in single_step_moves(board, die):
        nxt = apply_normalized(mv, board)
        _explore(nxt, rest, acc + [mv], out, max_used)


def _flip(x: int) -> int:
    return x if (x == 0 or x == 25) else 25 - x


def _denormalize(play: Play, player: int) -> Play:
    if player == WHITE:
        return play
    moves = tuple(CheckerMove(_flip(m.frm), _flip(m.to), m.die) for m in play.moves)
    return Play(moves, play.result.normalized(player))


def legal_plays(board: Board, d1: int, d2: int, player: int) -> List[Play]:
    norm = board.normalized(player)
    is_double = d1 == d2
    orderings = [[d1, d1, d1, d1]] if is_double else [[d1, d2], [d2, d1]]
    out: dict = {}
    max_used = [0]
    for seq in orderings:
        _explore(norm, seq, [], out, max_used)
    best = [p for p in out.values() if len(p.moves) == max_used[0]]
    if max_used[0] == 1 and not is_double:
        larger = max(d1, d2)
        with_larger = [p for p in best if p.moves[0].die == larger]
        if with_larger:
            best = with_larger
    if not best:
        chosen = [Play(tuple(), norm)]
    else:
        chosen = best
    return [_denormalize(p, player) for p in chosen]

File: web/engine.js
// engine.js — backgammon rules engine (browser + Node).
// A faithful port of Sources/BackgammonEngine (validated by dev/oracle).
//
// Conventions: points[1..24] from White's view, index 0 unused.
//   +n = n White checkers, -n = n Black checkers. White (= +1) moves high->low
//   and bears off past 1; Black (= -1) mirrors. All generation runs on a
//   normalized board (mover positive) and is de-normalized on the way out.
(function (root, factory) {
  const mod = factory();
  if (typeof module === "object" && module.exports) module.exports = mod;
  else root.Engine = mod;
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  const WHITE = 1, BLACK = -1;

  function startingBoard() {
    const pts = new Array(25).fill(0);
    const setup = { 1: -2, 6: 5, 8: 3, 12: -5, 13: 5, 17: -3, 19: -5, 24: 2 };
    for (const k in setup) pts[+k] = setup[k];
    return { points: pts, barW: 0, barB: 0, offW: 0, offB: 0 };
  }

  const clone = (b) => ({ points: b.points.slice(), barW: b.barW, barB: b.barB, offW: b.offW, offB: b.offB });
  const bar = (b, pl) => (pl === WHITE ? b.barW : b.barB);
  const off = (b, pl) => (pl === WHITE ? b.offW : b.offB);
  const owns = (b, p, pl) => (pl === WHITE ? b.points[p] > 0 : b.points[p] < 0);
  const count = (b, p, pl) => { const v = b.points[p]; return pl === WHITE ? (v > 0 ? v : 0) : (v < 0 ? -v : 0); };

  function checkerCount(b, pl) {
    let c = bar(b, pl) + off(b, pl);
    for (let p = 1; p <= 24; p++) c += count(b, p, pl);
    return c;
  }

  function mirrored(b) {
    const np = new Array(25).fill(0);
    for (let p = 1; p <= 24; p++) np[p] = -b.points[25 - p];
    return { points: np, barW: b.barB, barB: b.barW, offW: b.offB, offB: b.offW };
  }
  const normalized = (b, pl) => (pl === WHITE ? b : mirrored(b));

  function winner(b) { if (b.offW === 15) return WHITE; if (b.offB === 15) return BLACK; return null; }

  function winType(b) {
    const w = winner(b);
    if (w === null) return null;
    const loser = -w;
    if (off(b, loser) > 0) return "single";
    if (bar(b, loser) > 0) return "backgammon";
    const home = w === WHITE ? [1, 6] : [19, 24];
    for (let p = home[0]; p <= home[1]; p++) if (owns(b, p, loser)) return "backgammon";
    return "gammon";
  }

  function pipCount(b, pl) {
    let pips = bar(b, pl) * 25;
    for (let p = 1; p <= 24; p++) if (owns(b, p, pl)) {
      const dist = pl === WHITE ? p : 25 - p;
      pips += count(b, p, pl) * dist;
    }
    return pips;
  }

  // --- normalized helpers (mover positive) ---
  const canLand = (b, p) => b.points[p] >= -1;
  function allHome(b) {
    if (b.barW > 0) return false;
    for (let p = 7; p <= 24; p++) if (b.points[p] > 0) return false;
    return true;
  }
  function noneHigher(b, p) {
    if (p >= 6) return true;
    for (let q = p + 1; q <= 6; q++) if (b.points[q] > 0) return false;
    return true;
  }

  function singleStepMoves(b, die) {
    const moves = [];
    if (b.barW > 0) {
      const entry = 25 - die;
      if (canLand(b, entry)) moves.push({ from: 25, to: entry, die });
      return moves;
    }
    for (let p = 24; p >= 1; p--) if (b.points[p] > 0) {
      const dest = p - die;
      if (dest >= 1 && canLand(b, dest)) moves.push({ from: p, to: dest, die });
    }
    if (allHome(b)) {
      for (let p = 1; p <= 6; p++) if (b.points[p] > 0) {
        if (p === die) moves.push({ from: p, to: 0, die });
        else if (die > p && noneHigher(b, p)) moves.push({ from: p, to: 0, die });
      }
    }
    return moves;
  }

  // Apply a single move on a NORMALIZED board (mover positive).
  function applyNorm(m, b) {
    const nb = clone(b);
    if (m.from === 25) nb.barW--; else nb.points[m.from]--;
    if (m.to === 0) nb.offW++;
    else {
      if (nb.points[m.to] < 0) { nb.points[m.to] = 0; nb.barB++; }
      nb.points[m.to]++;
    }
    return nb;
  }

  // Apply a single move on a REAL board for `pl` (used by UI/AI step-by-step).
  function applyMove(b, m, pl) {
    const nb = clone(b);
    if (m.from === 25) { if (pl === WHITE) nb.barW--; else nb.barB--; }
    else nb.points[m.from] -= pl;
    if (m.to === 0) { if (pl === WHITE) nb.offW++; else nb.offB++; }
    else {
      if (nb.points[m.to] === -pl) { nb.points[m.to] = 0; if (pl === WHITE) nb.barB++; else nb.barW++; }
      nb.points[m.to] += pl;
    }
    return nb;
  }

  const keyB = (b) => b.points.join(",") + "|" + b.barW + "," + b.barB + "|" + b.offW + "," + b.offB;

  function explore(b, remaining, acc, out, state) {
    if (acc.length > state.max) state.max = acc.length;
    if (acc.length > 0) {
      const k = acc.map((m) => m.from + ">" + m.to + ":" + m.die).join(";") + "#" + keyB(b);
      out.set(k, { moves: acc.slice(), result: b });
    }
    if (remaining.length === 0) return;
    const die = remaining[0], rest = remaining.slice(1);
    for (const mv of singleStepMoves(b, die)) explore(applyNorm(mv, b), rest, acc.concat([mv]), out, state);
  }

  const flip = (x) => (x === 0 || x === 25 ? x : 25 - x);
  function denorm(play, pl) {
    if (pl === WHITE) return play;
    const moves = play.moves.map((m) => ({ from: flip(m.from), to: flip(m.to), die: m.die }));
    return { moves, result: normalized(play.result, pl) };
  }

  function legalPlays(b, d1, d2, pl) {
    const norm = normalized(b, pl);
    const isDouble = d1 === d2;
    const orderings = isDouble ? [[d1, d1, d1, d1]] : [[d1, d2], [d2, d1]];
    const out = new Map();
    const state = { max: 0 };
    for (const seq of orderings) explore(norm, seq, [], out, state);
    let best = [...out.values()].filter((p) => p.moves.length === state.max);
    if (state.max === 1 && !isDouble) {
      const larger = Math.max(d1, d2);
      const wl = best.filter((p) => p.moves[0].die === larger);
      if (wl.length) best = wl;
    }
    if (best.length === 0) best = [{ moves: [], result: norm }];
    return best.map((p) => denorm(p, pl));
  }

  // Distinct legal plays by *resulting board* (collapses the many move orders
  // that reach the same position — essential for fast search on doubles).
  function uniquePlays(b, d1, d2, pl) {
    const all = legalPlays(b, d1, d2, pl);
    const seen = new Map();
    for (const p of all) { const k = keyB(p.result); if (!seen.has(k)) seen.set(k, p); }
    return [...seen.values()];
  }

  // All 21 distinct rolls with their /36 weight.
  const allRolls = (() => {
    const r = [];
    for (let a = 1; a <= 6; a++) for (let b = a; b <= 6; b++) r.push({ d1: a, d2: b, prob: a === b ? 1 : 2 });
    return r;
  })();

  return {
    WHITE, BLACK,
    startingBoard, clone, bar, off, owns, count, checkerCount,
    mirrored, normalized, winner, winType, pipCount,
    singleStepMoves, applyNorm, applyMove, legalPlays, uniquePlays, keyB, allRolls,
  };
});

